Use batch requests when possible

kangzj opened the pull request:
## Proposed changes:

This PR implements request batching for analytics tracking pixels to improve performance by sending multiple tracking events together rather than individually.

### Changes:
* Queue tracking pixels during request execution instead of sending immediately
* Send all queued pixels together on the shutdown hook using WordPress Requests library
* Use `Requests::request_multiple()` with `blocking => false` for parallel non-blocking execution when available
* Fallback to sequential non-blocking `wp_remote_get()` calls when batch API is unavailable
* Maintains full backward compatibility with existing `WC_Tracks_Client` implementation

### Benefits:
* **Performance improvement**: Parallel execution reduces total request time (e.g., 10 pixels execute in ~100ms vs ~1000ms sequential)
* **Non-blocking**: Uses `blocking => false` to ensure zero impact on page load time
* **Efficient batching**: Reduces overhead by grouping related tracking requests
* **WordPress-native**: Uses WordPress Requests library instead of direct curl usage
* **Graceful degradation**: Automatically falls back when batching is unavailable

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- p2 discussion link if applicable -->

## Does this pull request change what data or activity we track or use?

No, this PR does not change what data is tracked. It only changes *how* tracking pixels are sent (batched vs individual requests). The same events with the same data are still being tracked.

## Testing instructions:

### Prerequisites:
* Set up a WooCommerce site with Jetpack and WooCommerce Analytics enabled
* Ensure analytics consent is granted (required for events to fire)

### Test 1: Verify events still fire correctly
1. Navigate to a WooCommerce store page that triggers multiple analytics events (e.g., product page, cart page)
2. Open browser developer tools → Network tab
3. Filter for requests to `/wp-json/woocommerce-analytics/v1/track`
4. The response time should be ~1/2 than it used to be for two+ events
5. Check that events appear in ClickHouse

Note: on my local, this reduced response time from 500ms+ to ~250ms for two events.

kangzj committed:
WooCommerce Analytics: Implement batched pixel requests

Implements request batching for analytics tracking pixels to improve performance by sending multiple tracking events together rather than individually.

Changes:
- Queue tracking pixels during request and send on shutdown hook
- Use WordPress Requests library request_multiple() for parallel execution when available (100ms timeout)
- Fallback to sequential non-blocking wp_remote_get() when batch API unavailable
- Maintains backward compatibility with existing WC_Tracks_Client implementation

Benefits:
- Parallel execution reduces total request time (10 pixels ~100ms vs ~1000ms sequential)
- Non-blocking fallback ensures no impact when batching unavailable
- Minimal overhead with only 2 new class properties

kangzj committed:
Add changelog

kangzj committed:
Fix build_pixel_url() method call - remove incorrect parameter

Nikschavan commented on projects/packages/woocommerce-analytics/src/class-wc-analytics-tracking.php:
```diff
@@ -152,7 +166,131 @@ private static function record_ch_event( $properties ) {
+
+		// Register shutdown hook once.
+		if ( ! self::$shutdown_hook_registered ) {
+			add_action( 'shutdown', array( __CLASS__, 'send_batched_pixels' ), 1 );
```
Should the priority be higher instead? If there is any important work scheduled to trigger on shutdown, ideally I imagine external requests should run before that as if there is any possible delay (Maybe in non batch supporting method), that will not affect anything else on the site?

kangzj replied:
I think it should be fine because the endpoint is tailored for dealing with logs specifically, and only loads three plugins. I don't imagine there will be more important tasks than sending the logs.

Curious to know what @chihsuan thinks.

chihsuan replied:
I agree that @Nikschavan's suggestions make sense. In ideal situations, this approach should work well. However, there are cases where the speed module may not be active. For example, if it fails to be copied to the `mu-plugins` folder. Moreover, this class handles not just endpoint requests, but also handles server-side events triggered by hooks.

I've also checked the WooCommerce Core and noticed that the `shutdown` hook is used to clear transients, save session data, and manage storage or flush notices.

That said, this likely doesn’t have a significant effect now, since methods that do not support batching are executed immediately rather than within the shutdown hook.

kangzj replied:
>  if it fails to be copied to the mu-plugins folder. 

@chihsuan That's a good point - I'm happy to accommodate any changes, but just wanted to know more about it. Are we saying, we should change the priority number to higher (actually it's lower priority)? Like 10 or something?

kangzj replied:
Changed priority to `20`

Nikschavan reviewed (dismissed):
Code looks good, Added a non-blocking suggestion

chihsuan commented on projects/packages/woocommerce-analytics/src/class-wc-analytics-tracking.php:
```diff
@@ -152,7 +166,131 @@ private static function record_ch_event( $properties ) {
+		$can_batch = ( class_exists( 'WpOrg\Requests\Requests' ) && method_exists( 'WpOrg\Requests\Requests', 'request_multiple' ) )
+			|| ( class_exists( 'Requests' ) && method_exists( 'Requests', 'request_multiple' ) );
+
+		if ( $can_batch ) {
```
Would it make sense to check `can_batch` before adding to the pixel queue? For example, if `can_batch` is true, we add to the pixel queue; otherwise, we record the event as before.

kangzj replied:
Addressed

chihsuan reviewed (commented):
> Note: on my local, this reduced response time from 500ms+ to ~250ms for two events.

Amazing! 👍 Thanks for working on this.

It might be worthwhile to contribute this back to WooCommerce core as a follow-up.

kangzj committed:
WooCommerce Analytics: Only use batched requests when request_multiple is available

Optimize batching logic to check for request_multiple support before queuing events. When batching is not supported, fall back to the original implementation by sending events immediately using WC_Tracks_Client::record_event(). This ensures events are not unnecessarily queued when they cannot be batched, improving reliability on systems without the Requests library batching capability.

kangzj committed:
WooCommerce Analytics: Remove unused send_with_individual_requests method

Remove the send_with_individual_requests method as it is no longer needed. With the updated logic, non-batching environments now send events immediately using WC_Tracks_Client::record_event() instead of queuing them.

kangzj committed:
Fix type mismatch: send pixel URL directly for non-batching fallback

Replace WC_Tracks_Client::record_event() call with direct wp_remote_get() to send the pixel URL. WC_Tracks_Client::record_event() expects an array parameter but we have an event object, and we've already built the pixel URL. This matches the original implementation approach of sending pixel URLs via HTTP GET requests.

kangzj committed:
Use WC_Tracks_Client::record_event for non-batching fallback

Simplify the non-batching fallback to use WC_Tracks_Client::record_event() directly, matching the existing pattern used elsewhere in the codebase. Add Phan suppression annotation to handle the type mismatch warning, as WC_Tracks_Client::record_event() accepts both array and event object parameters at runtime.

kangzj committed:
Add phpcs ignore comment before Phan suppression

Ensure proper comment formatting by including phpcs ignore before the Phan suppression annotation.

kangzj committed:
Fix Phan type mismatch suppression for WC_Tracks_Client::record_event

Corrected the Phan suppression format from @phan-suppress-current-line to @phan-suppress-next-line and used the correct error type PhanTypeMismatchArgument instead of PhanTypeMismatchArgumentProbablyReal. This properly suppresses the type mismatch error when passing WC_Tracks_Event or WC_Analytics_Ch_Event objects to WC_Tracks_Client::record_event() which expects an array.

kangzj committed:
Remove error_log call from batch pixel request failure handler

Silently fail without logging when batch pixel requests fail, as tracking pixels should not generate error logs or break the site.

kangzj committed:
Use WooCommerce logger instead of error_log for batch pixel failures

Replace error_log with wc_get_logger() to properly log batch pixel request failures through WooCommerce's logging system. This provides better integration with WooCommerce's status tools and maintains consistency with the rest of the WooCommerce ecosystem.

chihsuan reviewed (dismissed):
Thanks for the changes! @kangzj  LGTM 🙌

kangzj commented on projects/packages/woocommerce-analytics/src/class-wc-analytics-tracking.php:
```diff
@@ -152,7 +166,112 @@ private static function record_ch_event( $properties ) {
+
+		// Register shutdown hook once.
+		if ( ! self::$shutdown_hook_registered ) {
+			add_action( 'shutdown', array( __CLASS__, 'send_batched_pixels' ), 1 );
```
```suggestion
			add_action( 'shutdown', array( __CLASS__, 'send_batched_pixels' ), 20 );
```

kangzj committed:
Apply suggestion from @kangzj

kangzj commented:
Merging... will open up a PR to add batching to WooCommerce... Also will follow up with the job priority if 20 doesn't feel like the right one 😄

chihsuan commented:
> Merging... will open up a PR to add batching to WooCommerce... Also will follow up with the job priority if 20 doesn't feel like the right one 😄

Thanks, 20 sounds like a good priority to me. 👍
